module/zstd/zstd.c

```diff
@@ -0,0 +1,626 @@
+		printk(KERN_INFO "zstd: invalid compressed buffer size\n");
+#endif
+		return (1);
+	}
+
+	/*
+	 * Returns 0 on success (decompression function returned non-negative)
+	 * and non-zero on failure (decompression function returned negative.
+	 */
+	size_t len = real_zstd_decompress(
+	    &src[sizeof (bufsiz) + sizeof (zstdlevel)], \
+	    d_start, bufsiz, d_len);
+	if (ZSTD_isError(len)) {
+#ifdef _KERNEL
+		printk(KERN_INFO "zstd: decompression failed %ld\n", len);
+#endif
+		return (1);
+	}
+
+	return (0);
+}
+
+int
+zstd_getlevel(void *s_start, size_t s_len __unused)
+{
+	const char *src = s_start;
+	uint32_t cookie = BE_IN32(&src[sizeof (uint32_t)]);
+	int32_t zstdlevel = zstd_level_to_enum(cookie);
+
+	ASSERT(zstdlevel > ZIO_ZSTDLVL_INHERIT);
+	ASSERT(zstdlevel < ZIO_ZSTDLVL_LEVELS);
+
+	return (zstdlevel);
+}
+
+static size_t
+real_zstd_compress(const char *source, char *dest, int isize, int osize,
+    int level)
+{
+	size_t result;
+	ZSTD_CCtx *cctx;
+
+	ASSERT(level != 0);
+	if (level == ZIO_ZSTDLVL_DEFAULT)
+		level = ZIO_ZSTD_LEVEL_DEFAULT;
+
+	cctx = ZSTD_createCCtx_advanced(zstd_malloc);
+	/*
+	 * out of kernel memory, gently fall through - this will disable
+	 * compression in zio_compress_data
+	 */
+	if (cctx == NULL)
+		return (0);
+
+	result = ZSTD_compressCCtx(cctx, dest, osize, source, isize, level);
+
+	ZSTD_freeCCtx(cctx);
+	return (result);
+}
+
+
+static size_t
+real_zstd_decompress(const char *source, char *dest, int isize, int maxosize)
+{
+	size_t result;
+	ZSTD_DCtx *dctx;
+
+	dctx = ZSTD_createDCtx_advanced(zstd_malloc);
+	if (dctx == NULL)
+		return (ZSTD_error_memory_allocation);
+
+	result = ZSTD_decompressDCtx(dctx, dest, maxosize, source, isize);
+
+	ZSTD_freeDCtx(dctx);
+	return (result);
+}
+static int zstd_meminit(void);
+
+
+
+extern void *
+zstd_alloc(void *opaque __unused, size_t size)
+{
+	size_t nbytes = sizeof (struct zstd_kmem) + size;
+	struct zstd_kmem *z = NULL;
+	enum zstd_kmem_type type;
+	enum zstd_kmem_type newtype;
+	int i;
+	type = ZSTD_KMEM_UNKNOWN;
+	for (i = 0; i < ZSTD_KMEM_COUNT; i++) {
+		if (nbytes <= zstd_cache_size[i].kmem_size) {
+			type = zstd_cache_size[i].kmem_type;
+			if (zstd_kmem_cache[type]) {
+				z = kmem_cache_alloc( \
+				    zstd_kmem_cache[type], \
+				    zstd_cache_size[i].kmem_flags);
+				if (z) {
+					memset(z, 0, nbytes);
+					z->isvm = B_FALSE;
+				}
+			}
+			break;
+		}
+	}
+	newtype = type;
+	/* No matching cache */
+	if (type == ZSTD_KMEM_UNKNOWN || z == NULL) {
+		/*
+		 * consider max allocation size
+		 * so we need to use standard vmem allocator
+		 */
+#ifdef _KERNEL
+		z = vmem_zalloc(nbytes, KM_SLEEP);
+#else
+		z = kmem_zalloc(nbytes, KM_SLEEP);
+#endif
+		if (z)
+			newtype = ZSTD_KMEM_UNKNOWN;
+	}
+	/* fallback if everything fails */
+	if (!z && zstd_vmem_cache[type].vm && type == ZSTD_KMEM_DCTX) {
+		mutex_enter(&zstd_vmem_cache[type].barrier);
+		mutex_exit(&zstd_vmem_cache[type].barrier);
+
+		mutex_enter(&zstd_vmem_cache[type].barrier);
```

Comment:
Is this a merge conflict? It seems really strange to enter, then immediately exit, then re-enter this mutex. Also, it seems strange to only allow a single allocation with this mutex, why is that done?

Reply:
no the first waits for the barrier. the second enters the next.

Reply:
this is a special fallback code for decompression which is not thread safe. since i just have one reserved block. so i create this barrier to ensure that the memory reserved for decompression fallback is used only in a single thread

Reply:
i also removed the assert now. its no a requirement. so if someone with a 64 mbyte bochs is trying to emulate a early 90s pc with zfs and zstd, it will not crash anymore